### PR TITLE
[SPARK-30189][SQL] Interval from year-month/date-time string should handle whitespaces

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Currently, we parse interval from multi units strings or from date-time/year-month pattern strings, the former handles all whitespace, the latter not or even spaces.

### Why are the changes needed?

behavior consistency

### Does this PR introduce any user-facing change?
yes, interval in date-time/year-month like
```
select interval '\n-\t10\t 12:34:46.789\t' day to second
-- !query 126 schema
struct<INTERVAL '-10 days -12 hours -34 minutes -46.789 seconds':interval>
-- !query 126 output
-10 days -12 hours -34 minutes -46.789 seconds
``` 
is valid now.

### How was this patch tested?

add ut.

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu, thanks in advance.

**Comment by maropu** [comment]
also cc: @MaxGekk

**Comment by maropu** [comment]
Just in case, have you checked the changes of performance numbers in `IntervalBenchmark`? No change?

**Comment by MaxGekk** [comment]
@yaooqinn Maybe I have missed something but where does the requirement of supporting `\n`, `\r` ( and anything behind of ` `) come from?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 1]
```diff
@@ -156,7 +155,7 @@ object IntervalUtils {
   }
 
   private val dayTimePattern =
-    "^([+|-])?((\\d+) )?((\\d+):)?(\\d+):(\\d+)(\\.(\\d+))?$".r
+    "^([+|-](\\s+)?)?((\\d+)\\s+)?((\\d+):)?(\\d+):(\\d+)(\\.(\\d+))?$".r
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: Is `(\\s+)?` equal to `\\s*`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 2]
```diff
@@ -113,7 +113,7 @@ object IntervalUtils {
     result
   }
 
-  private val yearMonthPattern = "^([+|-])?(\\d+)-(\\d+)$".r
+  private val yearMonthPattern = "^([+|-])?(\\s+)?(\\d+)-(\\d+)$".r
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: Why not just `\\s*`?

**Comment by MaxGekk** [comment]
Before patching the buggy method, can we wait till this https://github.com/apache/spark/pull/26473 be merged?

**Comment by yaooqinn** [comment]
> Just in case, have you checked the changes of performance numbers in `IntervalBenchmark`? No change?

AFAIK，`IntervalBenchmark` only test multi unit interval strings

**Comment by yaooqinn** [comment]
> @yaooqinn Maybe I have missed something but where does the requirement of supporting `\n`, `\r` ( and anything behind of ` `) come from?

We have support dealing with ISO whitespaces for almost all internal types

**Comment by cloud-fan** [comment]
is this still valid?

**Comment by yaooqinn** [comment]
I will check this fix as ASAP, thanks for reminding me.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala [hunk 3]
```diff
@@ -312,6 +321,11 @@ class IntervalUtilsSuite extends SparkFunSuite with SQLHelper {
     checkFail("5 30:12:20", DAY, SECOND, "hour 30 outside range")
     checkFail("5 30-12", DAY, SECOND, "must match day-time format")
     checkFail("5 1:12:20", HOUR, MICROSECOND, "Cannot support (interval")
+
+    // whitespaces
+    check("\t +5 12:40\t ", DAY, MINUTE, "5 days 12 hours 40 minutes")
+    checkFail("+5\t 12:40", DAY, MINUTE, "must match day-time format")
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: for multi-unit syntax, do we support `1 day \t 2 hours`?

Comment (yaooqinn) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/IntervalUtilsSuite.scala]: yes. My opinion is that the interval string in multi-unit syntax, we parse each value and each unit separately and they can be produced by the providers separately too so the whitespaces can be in the middle of each part. In year-month and day-time syntax, the string should be considered as a whole part so the whitespaces only appear on both ends, not in the middle.

take java.sql.Timestamp for an example, the timestamp string is treated as a whole

```
scala> java.sql.Timestamp.valueOf("2011-11-11 11:11:11.11\t")
res8: java.sql.Timestamp = 2011-11-11 11:11:11.11

scala> java.sql.Timestamp.valueOf("2011-11-11 \t 11:11:11.11\t")
java.lang.NumberFormatException: For input string: "	 11"
  at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
  at java.lang.Integer.parseInt(Integer.java:569)
  at java.lang.Integer.parseInt(Integer.java:615)
  at java.sql.Timestamp.valueOf(Timestamp.java:243)
  ... 28 elided
```

Also, we do the same in Spark

```
spark-sql> select timestamp '2013-01-02 \t 00:00:00';
Error in query:
Cannot parse the TIMESTAMP value: 2013-01-02 	 00:00:00(line 1, pos 7)

== SQL ==
select timestamp '2013-01-02 \t 00:00:00'
-------^^^

spark-sql> select timestamp '2013-01-02 00:00:00 \t';
2013-01-02 00:00:00
```

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 4]
```diff
@@ -804,6 +804,51 @@ struct<INTERVAL '1 days':interval>
...
+
+-- !query
+select interval '\n-\t10\t 12:34:46.789\t' day to second
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.sql.catalyst.parser.ParseException
+
+requirement failed: Interval string must match day-time format of '^(?<sign>[+|-])?(?<day>\d+) (?<hour>\d{1,2}):(?<minute>\d{1,2}):(?<second>(\d{1,2})(\.(\d{1,9}))?)$': 
+-	10	 12:34:46.789	, set spark.sql.legacy.fromDayTimeString.enabled to true to restore the behavior before Spark 3.0.(line 1, pos 16)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out]: not related to this PR, but we should only ask users tob set the legacy config if it works.

File: sql/core/src/test/resources/sql-tests/results/interval.sql.out [hunk 5]
```diff
@@ -1,5 +1,5 @@
 -- Automatically generated by SQLQueryTestSuite
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/interval.sql.out]: not related to this PR: we should put the test queries that have different behaviors if ansi mode is on to a different file, and only test that file with ansi mode on and off.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Pull request closed without merging**